packages/bridge/contracts/l1/L1DaiGateway.sol

```diff
@@ -81,51 +100,99 @@ contract L1DaiGateway is L1ArbitrumExtendedGateway {
-    address _from,
-    uint256 _tokenAmount,
-    uint256 _maxGas,
-    uint256 _gasPriceBid,
-    uint256 _maxSubmissionCost,
-    bytes memory _outboundCalldata
-  ) internal override returns (uint256) {
+  function outboundTransfer(
+    address l1Token,
+    address to,
+    uint256 amount,
+    uint256 maxGas,
+    uint256 gasPriceBid,
+    bytes calldata data
+  ) external payable returns (bytes memory) {
     // do not allow initiating new xchain messages if bridge is closed
     require(isOpen == 1, "L1DaiGateway/closed");
-
-    return
-      sendTxToL2(
-        inbox,
-        counterpartGateway,
-        _from,
-        msg.value, // we forward the L1 call value to the inbox
-        0, // l2 call value 0 by default
-        L2GasParams({
-          _maxSubmissionCost: _maxSubmissionCost,
-          _maxGas: _maxGas,
-          _gasPriceBid: _gasPriceBid
-        }),
-        _outboundCalldata
+    require(l1Token == l1Dai, "L1DaiGateway/token-not-dai");
+
+    // we use nested scope to avoid stack too deep errors
+    address from;
+    uint256 seqNum;
+    bytes memory extraData;
+    {
+      uint256 maxSubmissionCost;
+      (from, maxSubmissionCost, extraData) = parseOutboundData(data);
+      require(extraData.length == 0, "L1DaiGateway/call-hook-data-not-allowed");
+
+      TokenLike(l1Token).transferFrom(from, l1Escrow, amount);
+
+      bytes memory outboundCalldata = getOutboundCalldata(l1Token, from, to, amount, extraData);
+      seqNum = sendTxToL2(
+        l2Counterpart,
+        from,
+        0,
+        maxSubmissionCost,
+        maxGas,
+        gasPriceBid,
+        outboundCalldata
       );
+    }
+
+    emit DepositInitiated(l1Token, from, to, seqNum, amount);
+
+    return abi.encode(seqNum);
   }
 
-  function outboundEscrowTransfer(
-    address _l1Token,
-    address _from,
-    uint256 _amount
-  ) internal virtual override {
-    TokenLike(_l1Token).transferFrom(_from, l1Escrow, _amount);
+  function getOutboundCalldata(
+    address l1Token,
+    address from,
+    address to,
+    uint256 amount,
+    bytes memory data
+  ) public view returns (bytes memory outboundCalldata) {
+    bytes memory emptyBytes = "";
+
+    outboundCalldata = abi.encodeWithSelector(
+      ITokenGateway.finalizeInboundTransfer.selector,
+      l1Token,
+      from,
+      to,
+      amount,
+      abi.encode(emptyBytes, data)
+    );
+
+    return outboundCalldata;
   }
 
-  function inboundEscrowTransfer(
-    address _l1Token,
-    address _dest,
-    uint256 _amount
-  ) internal virtual override {
-    TokenLike(_l1Token).transferFrom(l1Escrow, _dest, _amount);
+  function finalizeInboundTransfer(
+    address l1Token,
+    address from,
+    address to,
+    uint256 amount,
+    bytes calldata data
+  ) external payable onlyL2Counterpart(l2Counterpart) {
+    require(l1Token == l1Dai, "L1DaiGateway/token-not-dai");
+    (uint256 exitNum, bytes memory callHookData) = abi.decode(data, (uint256, bytes));
```

Comment:
I would use a utility library to handle decoding. Makes it easier to keep track and maintain